toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -321,79 +321,81 @@ func DetachLoopbackDevice(diskDevPath string) (err error) {
+		partedArgument, err := partitionTableType.ConvertToPartedArgument()
 		if err != nil {
-			logger.Log.Warnf("Failed to create single partition")
+			logger.Log.Errorf("Unable to convert partition table type (%v) to parted argument", partitionTableType)
 			return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
 		}
-
-		partFsType, err := FormatSinglePartition(partDevPath, partition)
+		_, stderr, err = shell.Execute("flock", "--timeout", timeoutInSeconds, diskDevPath, "parted", diskDevPath, "--script", "mklabel", partedArgument)
 		if err != nil {
-			logger.Log.Warnf("Failed to format partition")
+			logger.Log.Warnf("Failed to set partition table type using parted: %v", stderr)
 			return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
 		}
 
-		if rootEncryption.Enable && partition.HasFlag(configuration.PartitionFlagDeviceMapperRoot) {
-			encryptedRoot, err = encryptRootPartition(partDevPath, partition, rootEncryption)
+		usingExtendedPartition := (len(disks[i].Partitions) > maxPrimaryPartitionsForMBR) && (partitionTableType == configuration.PartitionTableTypeMbr)
+
+		// Partitions assumed to be defined in sorted order
+		for idx, partition := range disks[i].Partitions {
+			partType, partitionNumber := obtainPartitionDetail(idx, usingExtendedPartition)
+			// Insert an extended partition
+			if partType == extendedPartitionType {
+				err = createExtendedPartition(diskDevPath, partitionTableType.String(), disks[i].Partitions, partIDToFsTypeMap, partDevPathMap)
+				if err != nil {
+					return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
+				}
+
+				// Update partType and partitionNumber
+				partType = logicalPartitionType
+				partitionNumber = partitionNumber + 1
+			}
+
+			partDevPath, err := CreateSinglePartition(diskDevPath, partitionNumber, partitionTableType.String(), partition, partType)
 			if err != nil {
-				logger.Log.Warnf("Failed to initialize encrypted root")
+				logger.Log.Warnf("Failed to create single partition")
 				return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
 			}
-			partDevPathMap[partition.ID] = GetEncryptedRootVolMapping()
-		} else if readOnlyRootConfig.Enable && partition.HasFlag(configuration.PartitionFlagDeviceMapperRoot) {
-			readOnlyRoot, err = PrepReadOnlyDevice(partDevPath, partition, readOnlyRootConfig)
+
+			partFsType, err := FormatSinglePartition(partDevPath, partition)
 			if err != nil {
-				logger.Log.Warnf("Failed to initialize read only root")
+				logger.Log.Warnf("Failed to format partition")
 				return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
 			}
-			partDevPathMap[partition.ID] = readOnlyRoot.MappedDevice
-		} else {
-			partDevPathMap[partition.ID] = partDevPath
-		}
 
-		partIDToFsTypeMap[partition.ID] = partFsType
+			if rootEncryption.Enable && partition.HasFlag(configuration.PartitionFlagDeviceMapperRoot) {
```

Comment:
Could you double check in the validator function `checkDeviceMapperFlags()` that only one disk sets this for a given systemconfig? (I'm fairly sure it works as is)

Reply:
after looking at checkDeviceMapperFla() it seems like only one disk would be selected for a given systemconfig. Based off the functions the disk that gets selected would be the first in the list that has the root partition "/"

Reply:
Ok, that was my reading as well.